sn_nat_detection/src/main.rs

```diff
@@ -0,0 +1,400 @@
+// Unless required by applicable law or agreed to in writing, the SAFE Network Software distributed
+// under the GPL Licence is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied. Please review the Licences for the specific language governing
+// permissions and limitations relating to use of the SAFE Network Software.
+
+use clap::Parser;
+use futures::StreamExt;
+use libp2p::autonat::NatStatus;
+use libp2p::core::{multiaddr::Protocol, Multiaddr};
+use libp2p::swarm::SwarmEvent;
+use libp2p::{noise, tcp, yamux};
+use std::collections::HashSet;
+use std::error::Error;
+use std::net::Ipv4Addr;
+use std::time::Duration;
+use tracing::{debug, info, warn};
+use tracing_log::AsTrace;
+use tracing_subscriber::layer::SubscriberExt;
+use tracing_subscriber::util::SubscriberInitExt;
+use tracing_subscriber::EnvFilter;
+
+use behaviour::{Behaviour, BehaviourEvent};
+
+mod behaviour;
+
+const CONFIDENCE_MAX: usize = 2;
+const RETRY_INTERVAL: Duration = Duration::from_secs(10);
+
+/// A tool to detect NAT status of the machine. It can be run in server mode or client mode.
+/// The program will exit with an error code if NAT status is determined to be private.
+#[derive(Debug, Parser)]
+#[clap(name = "libp2p autonat")]
+struct Opt {
+    /// Port to listen on.
+    ///
+    /// `0` causes the OS to assign a random available port.
+    #[clap(long, short, default_value_t = 0)]
+    port: u16,
+
+    /// Servers to send dial-back requests to, in a 'multiaddr' format.
+    ///
+    /// A multiaddr looks like `/ip4/1.2.3.4/tcp/1200/tcp` where `1.2.3.4` is the IP and `1200` is the port.
+    /// Alternatively, the address can be written as `1.2.3.4:1200`.
+    ///
+    /// This argument can be provided multiple times to connect to multiple peers.
+    #[clap(name = "SERVER", value_name = "multiaddr", value_delimiter = ',', value_parser = parse_peer_addr)]
+    server_addr: Vec<Multiaddr>,
+
+    /// Disable use of UPnP to open a port on the router, before detecting NAT status.
+    #[clap(long, short, default_value_t = false)]
+    no_upnp: bool,
+
+    #[command(flatten)]
+    verbose: clap_verbosity_flag::Verbosity,
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn Error>> {
+    // Process command line arguments.
+    let opt = Opt::parse();
+
+    let registry = tracing_subscriber::registry().with(tracing_subscriber::fmt::layer());
+    // Use `RUST_LOG` if set, else use the verbosity flag (where `-vvvv` is trace level).
+    let _ = if std::env::var_os("RUST_LOG").is_some() {
+        registry.with(EnvFilter::from_env("RUST_LOG")).try_init()
+    } else {
+        let filter = tracing_subscriber::filter::Targets::new().with_target(
+            env!("CARGO_BIN_NAME").replace('-', "_"),
+            opt.verbose.log_level_filter().as_trace(),
+        );
+        registry.with(filter).try_init()
+    };
```

Comment:
Maybe in a follow up PR can we make use of `sn_logging` here? I think it will be useful to have the logs inside a single place.

Reply:
Yeah, kind of agree, though I think it's quite compact code here that isn't too complex. I didn't think it would be useful to log to files etc.